### PR TITLE
Add Linux SCHED_DEADLINE support

### PR DESCRIPTION
**Pull request by RafaelOrtizRC**
This adds rudimentary support for [`SCHED_DEADLINE`](https://www.kernel.org/doc/Documentation/scheduler/sched-deadline.txt) on Linux systems. It is tested and working on `Linux 5.4.0-84-generic #94-Ubuntu SMP Thu Aug 26 20:27:37 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`, but I see no reason it wouldn't work as far back as `3.14` when `SCHED_DEADLINE` was introduced.

There's a few ugly bits:
* Since `SCHED_DEADLINE` isn't POSIX, there's a bunch of `#[cfg(target_os = "linux")]` all over the place, a dedicated `linux.rs` might be better. I also noticed some `libc::` dependencies that aren't present on macOS, so I don't know that `unix.rs` is fully POSIX anyway.
* The `ThreadId` passed in is interpreted as a `pid_t` instead of being translated properly (solutions to which are also hacky), since the syscall for deadline scheduling expects a kernel tid (userspace pid).

### CODE REVIEW

**Comment by vityafx** [comment]
Thanks much for the contribution! Is there a way we can test this policy by any chance?

**Comment by RafaelOrtizRC** [comment]
> Thanks much for the contribution! Is there a way we can test this policy by any chance?

Sure! I added a test for the deadline policy as-implemented (and cleaned up some formatting stuff - CLion doesn't default to `rustfmt`).

**Comment by RafaelOrtizRC** [comment]
It looks like the `4.4` kernel on xenial (which is the GitHub Actions runner you're using) will refuse to admit any threads with deadline scheduling when runtime == deadline == period (even though that's supposed to be supported). I'm changing the interface to add a `ThreadPriority::Deadline(u64, u64, u64)` so the tests can specify values that the scheduler should consistently admit. Will probably push it up tomorrow.

**Comment by RafaelOrtizRC** [comment]
Sorry, I got pulled into other work. Anyway, the tests should be passing now. I added the explicit deadline interface as described and tested (working) on Xenial.

**Comment by RafaelOrtizRC** [comment]
Okay, _now_ it should be working. I removed the clippy warning in that one test (I think it's more clear with the identity multiplication in this specific case) and fixed the CI issues (hopefully).

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -56,9 +56,13 @@ jobs:
...
-      - name: Test all features
-        if: matrix.features == ''
-        run: cargo test --all-features
+      - name: Test all features (macOS)
+        if: matrix.features == '' && runner.os == 'macOS'
+        run: cargo test --all-features -- --skip set_deadline_policy
+
+      - name: Test all features (other)
+        if: matrix.features == '' && runner.os != 'macOS'
+        run: sudo -E $HOME/bin/cargo test --all-features
```
Comment (vityafx) [inline .github/workflows/ci.yml:65]: Out of curiosity, why was this necessary?

Comment (RafaelOrtizRC) [reply .github/workflows/ci.yml:65]: macOS doesn't implement `SCHED_DEADLINE` since it's not part of POSIX. The test will only pass on Linux. It looks like the rest of the tests are failing due to cargo being installed in a different location than I expected, and `sudo` not being part of the environment. The test on Linux needs to be executed as root.

Comment (RafaelOrtizRC) [reply .github/workflows/ci.yml:65]: I'm not familiar enough with GitHub Actions to know what the correct fix is here. I assumed that the run command would be executed as a regular shell command inside a VM, but it appears that was wrong.

Comment (vityafx) [reply .github/workflows/ci.yml:65]: Can you try changing it to `sudo -E /usr/share/rust/.cargo/bin/cargo test --all-features` please?

Comment (RafaelOrtizRC) [reply .github/workflows/ci.yml]: Try again now.

Comment (RafaelOrtizRC) [reply .github/workflows/ci.yml]: Looks like that (mostly) worked, I pushed up a typo fix. Sorry for the spam, I'm not super familiar with GitHub Actions.

File: src/unix.rs [hunk 2]
```diff
@@ -219,20 +281,54 @@ pub fn thread_schedule_policy() -> Result<ThreadSchedulePolicy, Error> {
...
+                    sched_deadline: deadline as u64,
+                    sched_period: period as u64,
+
+                    ..Default::default()
+                };
+                libc::syscall(
+                    libc::SYS_sched_setattr,
+                    tid,
+                    &sched_attr as *const _,
+                    // we are not setting SCHED_FLAG_RECLAIM nor SCHED_FLAG_DL_OVERRUN
```
Comment (vityafx) [inline src/unix.rs:322]: Just curious, why don't we set it? Is it possible to set it?

Comment (vityafx) [reply src/unix.rs:322]: I am not using such a policy anywhere, do you think it is worth adding these options to the interface so that others could configure them?

Comment (vityafx) [reply src/unix.rs:322]: Sure! Thanks.

**Comment by vityafx** [comment]
Enjoy https://crates.io/crates/thread-priority/0.3.0 and thanks much for the contribution!
